tools/bridge/bridge/transfer_fetcher.py

```diff
@@ -1,25 +1,135 @@
+        token_contract_address: str,
+        foreign_bridge_contract_address: str,
+        max_reorg_depth: int,
+        transfer_event_fetch_limit: int,
+    ):
+        token_contract_code = w3_foreign.eth.getCode(token_contract_address)
+
+        if not token_contract_code:
+            raise ValueError(
+                f"The given token contract address {token_contract_address} "
+                "does not point to a contract!"
+            )
+
+        if TRANSFER_EVENT_SIGNATURE_HASH not in token_contract_code:
+            raise ValueError(
+                f"The token contract at the given address {token_contract_address}"
+                "does not have the (correct) 'Transfer' event!"
+            )
+
+        foreign_bridge_contract_code = w3_foreign.eth.getCode(
+            foreign_bridge_contract_address
+        )
+
+        if not foreign_bridge_contract_code:
+            raise ValueError(
+                f"The given foreign bridge contract address {foreign_bridge_contract_address} "
+                "does not point to a contract!"
+            )
+
+        self._logger = logging.getLogger("bridge.transfer_fetcher")
+        self._w3_foreign = w3_foreign
+        self._queue = queue
+        self._token_contract = w3_foreign.eth.contract(
+            address=token_contract_address, abi=MINIMAL_TOKEN_ABI
+        )
+        self._foreign_bridge_contract_address = foreign_bridge_contract_address
+        self._max_reorg_depth = max_reorg_depth
+        self._transfer_event_fetch_limit = transfer_event_fetch_limit
+        self._last_fetched_block_number = (
+            -1
+        )  # Fetching starts always one block after the last one.
+
+    def _fetch_token_transfer_events_in_range(
+        self, from_block_number: int, to_block_number: int
+    ) -> None:
+        from_block_number = max(from_block_number, 0)
+        to_block_number = min(to_block_number, self._w3_foreign.eth.blockNumber)
+
+        self._logger.debug(
+            "Fetch transfer events from block {from_block_number} to {to_block_number}."
+        )
+
+        assert from_block_number <= to_block_number, (
+            f"Can not fetch from a negative range of blocks "
+            f"({from_block_number} to {to_block_number})"
+        )
+
+        transfer_event_filter = self._token_contract.events.Transfer.createFilter(
+            fromBlock=from_block_number,
+            toBlock=to_block_number,
+            argument_filters={"to": self._foreign_bridge_contract_address},
+        )
+
+        events = transfer_event_filter.get_all_entries()
+
+        self._logger.debug(f"Found {len(events)} events.")
+
+        for event in events:
+            self._queue.put(event)
+
+    def _fetch_token_transfer_events_not_seen(self) -> None:
+        self._logger.debug("Fetch new transfer events.")
+        self._logger.debug(
+            f"Last fetched block number: {self._last_fetched_block_number}."
+        )
+
+        head_block_number = max(
```

Comment:
I think `head` is misleading as it usually refers to the head of the chain. Maybe `fetch_until`?

Reply:
:heavy_check_mark: